nextcord/application_command.py

```diff
@@ -743,15 +743,43 @@ def from_callback(
             #  might be able to do something here.
             if option_class:
                 skip_counter = 1
-                # TODO: use typing.get_type_hints when 3.9 is standard
-                typehints = typing_extensions.get_type_hints(self.callback, include_extras=True)
                 # Getting the callback with `self_skip = inspect.ismethod(self.callback)` was problematic due to the
                 #  decorator going into effect before the class is instantiated, thus being a function at the time.
                 #  Try to look into fixing that in the future?
                 #  If self.parent_cog isn't reliable enough, we can possibly check if the first parameter name is `self`
                 if self.parent_cog:
                     skip_counter += 1
 
+                # TODO: use typing.get_type_hints when 3.9 is standard
+                typehints = typing_extensions.get_type_hints(self.callback, include_extras=True)
+                callback_params = signature(self.callback).parameters
+                possible_options: Set[str] = set()
+
+                for i, param in enumerate(callback_params.values()):
+                    if (
+                        # could be a self or interaction parameter
+                        param.annotation is not param.empty
+                        # will always be an interaction parameter
+                        and not issubclass(param.annotation, Interaction)
+                        # will always be a self parameter
+                        # TODO: use typing.Self when 3.11 is standard
+                        and param.annotation is not typing_extensions.Self
+                        # will always be a self parameter
+                        and not isinstance(param.annotation, typing.TypeVar)
+                    ) or i not in (0, 1):
+                        possible_options.add(param.name)
+
+                difference = {param.name for param in callback_params.values()} - possible_options
+
+                if self.parent_cog is not None and len(difference) < 2:
+                    raise ValueError(
+                        f"Callback {self.error_name} is missing the self and/or interaction parameters. Please double check your function definition."
+                    )
+                elif len(difference) < 1:
+                    raise ValueError(
+                        f"Callback {self.error_name} is missing the interaction parameter. Please double check your function definition."
+                    )
```

Comment:
Since `difference` is only used for its length, it may be more readable to instead calculate it as a number.

If `difference` is not calculated using set difference, `possible_options` also does not need to be a set.

By using this information, it should be possible to instead calculate the number of parameters that aren't necessarily options and just use a single number variable.

I'm thinking it could be simplified like this. This should be tested, but I believe this is equivalent and eliminates the need for the for-loop.

```py
callback_params = signature(self.callback).parameters

# Count the number of parameters that are not necessarily command options
# Note, this count may include parameters that are used as command options
num_non_option_params = sum(
    # could be a self or interaction parameter
    param.annotation is param.empty
    # will always be an interaction parameter
    or issubclass(param.annotation, Interaction)
    # will always be a self parameter
    # TODO: use typing.Self when 3.11 is standard
    or param.annotation is typing_extensions.Self
    # will always be a self parameter
    or isinstance(param.annotation, typing.TypeVar)
    for param in list(signature(self.callback).parameters.values())[:2]
)

if self.parent_cog is not None and num_non_option_params < 2:
    raise ValueError(
        f"Callback {self.error_name} is missing the self and/or interaction parameters. Please double check your function definition."
    )
elif num_non_option_params < 1:
    raise ValueError(
        f"Callback {self.error_name} is missing the interaction parameter. Please double check your function definition."
    )
```

Note: it may even be possible to use `[:skip_counter]` instead of `[:2]` since you probably don't need to check the first *two* parameters when it is not in a cog.